src/org/etools/j1939_84/controllers/part01/Part01Step12Controller.java

```diff
@@ -230,34 +230,47 @@ protected void run() throws Throwable {
 
-    private void getDm58AndVerifyData(int moduleAddress) {
+    private void getDm58AndVerifyData() {
 
         // 6.1.12.5 Actions3:
         // a. DS DM7 with TID 245 (for DM58) using FMI 31 for first SP identified as not supporting DM58 in a DM24
         // response In step 6.1.4.1 to the SP’s respective OBD ECU. (Use of an SP that supports test results is
         // preferred when available).
+        var obdModules = getDataRepository().getObdModules()
+                                            .stream()
+                                            .filter(module -> {
+                                                var supportedSPNs = module.getSupportedSPNs()
+                                                                          .stream()
+                                                                          .filter(supportedSPN -> !supportedSPN.supportsRationalityFaultData())
+                                                                          .collect(Collectors.toList());
+                                                if (!supportedSPNs.isEmpty()) {
+                                                    return true;
+                                                }
+                                                return false;
+                                            })
+                                            .collect(Collectors.toList());
+
         var nonRatFaultSps = getDataRepository().getObdModules()
                                                 .stream()
-                                                .filter(module -> module.getSourceAddress() == moduleAddress)
                                                 .flatMap(m -> m.getSupportedSPNs().stream())
                                                 .filter(supported -> !supported.supportsRationalityFaultData())
                                                 .collect(Collectors.toList());
-        if (nonRatFaultSps.isEmpty()) {
-            getListener().onResult("6.1.12.5.a - No SPs found that do NOT indicate support for DM58 in the DM24 response from "
-                    + Lookup.getAddressName(moduleAddress));
+        if (nonRatFaultSps.isEmpty() || obdModules.isEmpty()) {
+            getListener().onResult("6.1.12.5.a - No SPs found that do NOT indicate support for DM58 in the DM24 response");
         } else {
             int requestSpn = nonRatFaultSps.stream()
                                            .filter(SupportedSPN::supportsScaledTestResults)
                                            .findFirst()
                                            .orElseGet(() -> nonRatFaultSps.get(0))
                                            .getSpn();
 
-            var packet = getCommunicationsModule().requestDM58(getListener(), moduleAddress, requestSpn)
+            var packet = getCommunicationsModule().requestDM58(getListener(),
+                                                               obdModules.get(0).getSourceAddress(),
```

Comment:
The address you're sending to should be a variable to prevent it from changing here and not changing below.